src/yaml_runner/engines/base_engine.py

```diff
@@ -0,0 +1,136 @@
+
+    @property
+    def config(self):
+        """The config currently in use by the YamlRunner"""
+        return self._config.copy()
+
+    @config.setter
+    def config(self,config: dict):
+        if isinstance(config,dict):
+            self._config = config
+        else:
+            raise TypeError('Expected config as type: dict')
+
+    @abstractmethod
+    def get_commands(self, args: list[str]) -> list[str]:
+        """
+        This method takes in the cli args as a list of strings.
+        Setup the parsers. Parses the arguments to work out the commands to be run.
+        Returns the commands to be run as a list of strings.
+
+        Returns:
+            list[str]: List of strings, commands to be run.
+        """
+        pass
+
+    @abstractmethod
+    def _process_arguments(self, args: list[str]) -> dict:
+        """
+        Sets up and runs argument parser/s for each command section
+        found in the config. Returning the namespace object from the parsed
+        arguments as a dictionary.
+
+        Args:
+            args (list[str]): Unparsed arguments in a list.
+
+        Returns:
+            dict: Dictionary of arguments and their values.
+        """
+        pass
+
+    def _get_command_sections(self, parsed_config: dict) -> list[str]:
+        """
+        Recursive function to extract the command sections from a dictionary.
+        
+        Args:
+        parsed_config (dict): Dictionary containing command configiguration data.
+        
+        Returns:
+        A list of dictionaries containing command sections from the parsed configuration. Each
+        dictionary includes the 'name' key with the corresponding key from the parsed configuration.
+        """
+        command_dicts = []
+        for key, value in parsed_config.items():
+            if isinstance(value,dict):
+                if value.get('command',None):
+                    value.update({'name':key})
+                    command_dicts.append(value)
+                else:
+                    command_dicts += self._get_command_sections(value)
+        return command_dicts
+
+    def _add_params_parser(self,parser: argparse.ArgumentParser, params:dict):
+        """Add command parameters to the commands parser.
+        Takes in the commands "params" section from the config and the commands subparser.
+
+        Args:
+            parser (argparse.ArgumentParser): Commands subparser.
+            params (dict): Params section from yaml config.
+        """
+        if params.get('passthrough'):
+            parser.add_argument('passthrough',
+                                action='store',
+                                help='Extra arguments for the command.',
+                                nargs=argparse.REMAINDER,
+                                metavar='ARGUMENTS')
+
+    def _build_commands(self, commands:list[str], cli_args: dict) -> list[str]:
+        """_summary_
```

Comment:
need clearer docstring:
e.g. 'replaces any occurence of '$@' in teh command strings with the args provided by the users in the cli'